### PR TITLE
workflow stub injection

### LINKED ISSUE #38: Allow workflow stub injection
with the current implementation, using the workflow client is a bit tedious
```
@Inject
WorkflowClient client;

public void callWorkflow(){

        WorkflowOptions options = WorkflowOptions.newBuilder()
                .setTaskQueue("workerName")
                .setWorkflowId("money-transfer-workflow")
                .build();

        MoneyTransferWorkflow workflow = client.newWorkflowStub(MoneyTransferWorkflow.class, options);
        TransactionDetails transaction = new CoreTransactionDetails("249020073", "152354872", "57c65dea-e57e-4a0a", 68);
        workflow.transfer(transaction);
}
```

What we could do to simplify the process is making the workflow stub available as a CDI bean:

```
@Inject
@WorkflowStub(worker = "workerName", workflowId = "money-transfer-workflow")
MoneyTransferWorkflow workflow;

public void callWorkflow(){

        TransactionDetails transaction = new CoreTransactionDetails("249020073", "152354872", "57c65dea-e57e-4a0a", 68);
        workflow.transfer(transaction);
}
```

### PR DESCRIPTION
**Pull request by rmanibus**
fix #38

### CODE REVIEW

**Comment by rmanibus** [comment]
one of the test where two workflow implementations are present on two different queue is failing with the following error:

java.lang.RuntimeException: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#generateResources threw an exception: java.lang.IllegalStateException: java.util.concurrent.ExecutionException: java.lang.NullPointerException: Annotation class not 

```
available: @io.quarkiverse.temporal.WorkflowStub(worker = "namedWorker")
	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:858)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:256)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2513)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1538)
	at java.base/java.lang.Thread.run(Thread.java:1570)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
```

the following code is supposed to create an instance qualified by the worker name:

```
                            SyntheticBeanBuildItem.configure(workflowBuildItem.workflow)
                                    .scope(ApplicationScoped.class)
                                    .addQualifier()
                                    .annotation(WorkflowStub.class)
                                    .addValue("worker", worker)
                                    .done()
                                    .runtimeProxy(recorder.createWorkflowStub(clientBuildItem.workflowClient, workflowBuildItem.workflow, worker))
                                    .setRuntimeInit()
                                    .done()
```

Any idea about what I am doing wrong @melloware ?

**Comment by melloware** [comment]
Let me study the code and see if I see anything obvious.

**Comment by melloware** [comment]
From first glance this code looks correct to me.

However, the build is failing because "imports aren't sorted" make sure to run `mvn clean install` on the whole project to format it.

**Comment by rmanibus** [comment]
I think we should add the formatting as a pre-commit hook ! Or is there any good way to add it directly in intellij ?
Anyway I reformatted, and now it is failing with the NPE again

**Comment by melloware** [comment]
OK weird its saying the annotation is not available not that there are no matches?

```
Caused by: java.lang.NullPointerException: Annotation class not available: @io.quarkiverse.temporal.WorkflowStub(worker = "namedWorker")
	at java.base/java.util.Objects.requireNonNull(Objects.java:235)
	at io.quarkus.arc.processor.AnnotationLiteralProcessor.create(AnnotationLiteralProcessor.java:86)
```

I would have expected a different error.

**Comment by rmanibus** [comment]
https://github.com/quarkusio/quarkus/issues/18498

**Comment by melloware** [comment]
@gsmet any thoughts here?  This is sort of out of my depth.

**Comment by rmanibus** [comment]
Cc @mkouba It seems you found the culprit in the above issue, do you have any idea why this could be happening ?

**Comment by rmanibus** [comment]
Oups, we posted at the same time, didn't mean to bring everyone's focus on this

**Comment by rmanibus** [comment]
adding `io.quarkiverse.temporal.WorkflowStub.class` via addClass in the test fix the issue. Is there any buildStep to use to register the qualifier ? 

```
    @RegisterExtension
    static final QuarkusUnitTest unitTest = new QuarkusUnitTest()
            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
                    .addClass(SimpleWorkflow.class)
                    .addClass(DefaultSimpleWorkflowImpl.class)
                    .addClass(NamedSimpleWorkflowImpl.class)
                    .addClass(WorkflowStub.class)
                    .addAsResource(
                            new StringAsset("quarkus.temporal.start-workers: false\n"),
                            "application.properties"));
```

**Comment by rmanibus** [comment]
injecting the synthetic bean in the test also fix it, so this has to do with the bean existing but not being injected anywhere

```
    @Inject
    @WorkflowStub(worker = "namedWorker")
    SimpleWorkflow namedWorkflow;
```

**Comment by melloware** [comment]
here is what we do in MyFaces to register annotations.

```java
private static final String[] BEAN_DEFINING_ANNOTATION_CLASSES =
    {
            Named.class.getName(),
            FacesComponent.class.getName(),
            FacesBehavior.class.getName(),
            FacesConverter.class.getName(),
            FacesValidator.class.getName(),
            FacesRenderer.class.getName(),
            NamedEvent.class.getName(),
            FacesBehaviorRenderer.class.getName(),
            FaceletHandler.class.getName(),
            FlowDefinition.class.getName(),
            View.class.getName()
    };

@BuildStep
    void buildCdiBeans(BuildProducer<BeanDefiningAnnotationBuildItem> beanDefiningAnnotation)
    {

        for (String clazz : BEAN_DEFINING_ANNOTATION_CLASSES)
        {
            beanDefiningAnnotation.produce(new BeanDefiningAnnotationBuildItem(DotName.createSimple(clazz)));
        }

    }
```

**Comment by rmanibus** [comment]
I just tried, this is not fixing it. my understanding of the BeanDefiningAnnotationBuildItem is that it defines an alias for an existing scope

**Comment by melloware** [comment]
Got it. Yeah those are all existing already.

**Comment by rmanibus** [comment]
I believe this is a bug in quarkus

**Comment by melloware** [comment]
want me to ask on Zulip or are you on Zulip its a great way to ask questions of the devs

**Review by melloware — APPROVED** [review]

File: extension/deployment/src/test/java/io/quarkiverse/temporal/deployment/WorkflowStubMultipleInjectionTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClass(SimpleWorkflow.class)
+                    .addClass(UnnamedSimpleWorkflowImpl.class)
+                    .addClass(NamedSimpleWorkflowImpl.class)
+                    .addAsResource(
+                            new StringAsset("\n"),
+                            "application.properties"));
+
+    @Inject
+    WorkflowClient workflowClient;
```
Comment (rmanibus) [inline extension/deployment/src/test/java/io/quarkiverse/temporal/deployment/WorkflowStubMultipleInjectionTest.java:39]: @melloware another weirdness: if I remove this injection, the test fail with:
```
Caused by: java.lang.IllegalStateException: No bean type found for: PRODUCER METHOD bean [types=[io.temporal.client.WorkflowClient, java.lang.Object], qualifiers=[@Default, @Any], target=io.temporal.client.WorkflowClient client(), declaringBean=io.quarkiverse.temporal.deployment.WorkflowStubMultipleInjectionTest$Producers]
```

File: extension/deployment/src/test/java/io/quarkiverse/temporal/deployment/WorkflowStubMultipleInjectionTest.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+        WorkflowExecution otherExecution = WorkflowClient.start(defaultWorkflowWithOtherWorkflowId::transfer);
+        Assertions.assertEquals("otherWorkflowIdForSimpleWorkflow", otherExecution.getWorkflowId());
+
+    }
+
+    public static class Producers {
+        @Produces
+        WorkflowClient client() {
+            return testWorkflowEnvironment.getWorkflowClient();
+        }
```
Comment (rmanibus) [inline extension/deployment/src/test/java/io/quarkiverse/temporal/deployment/WorkflowStubMultipleInjectionTest.java:72]: is there any better way to provide an alternative implementation in a QuarkusUnitTest ?

File: extension/runtime/src/main/java/io/quarkiverse/temporal/WorkflowStub.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.Target;
+
+import jakarta.enterprise.util.Nonbinding;
+import jakarta.inject.Qualifier;
+
+@Qualifier
+@Retention(RUNTIME)
+@Target({ METHOD, FIELD, PARAMETER, TYPE })
+public @interface WorkflowStub {
```
Comment (rmanibus) [inline extension/runtime/src/main/java/io/quarkiverse/temporal/WorkflowStub.java:18]: Temporal already have an interface named `io.temporal.client.WorkflowStub`, maybe we should rename this ?

Comment (melloware) [reply extension/runtime/src/main/java/io/quarkiverse/temporal/WorkflowStub.java:18]: Yes we should rename this.  I am back to all of our custom Quarkus annotations should be `@TemporalXXX` this way we KNOW its ours and I know they didn't in Spring Boot but `@TemporalActivity` and `@TemporalWorkflowStub` are so much cleaner.

File: extension/deployment/src/test/java/io/quarkiverse/temporal/deployment/WorkflowStubMultipleInjectionTest.java (hunk 2 above)
Comment (melloware) [reply extension/deployment/src/test/java/io/quarkiverse/temporal/deployment/WorkflowStubMultipleInjectionTest.java:72]: Not that I am aware of.

File: extension/deployment/src/test/java/io/quarkiverse/temporal/deployment/WorkflowStubMultipleInjectionTest.java (hunk 1 above)
Comment (melloware) [reply extension/deployment/src/test/java/io/quarkiverse/temporal/deployment/WorkflowStubMultipleInjectionTest.java:39]: If you remove that injection then how will it get a WorkflowClient?

Comment (melloware) [reply extension/deployment/src/test/java/io/quarkiverse/temporal/deployment/WorkflowStubMultipleInjectionTest.java:39]: OK yeah this is bad then.  If you don't need the WorkflowClient here it shouldn't have to be injected so we have to somehow provide `@Alternative` bean or something that allows your code to work without this injection.

Comment (melloware) [reply extension/deployment/src/test/java/io/quarkiverse/temporal/deployment/WorkflowStubMultipleInjectionTest.java:39]: Or shouldn't a default one always be `produced` by this extension?

File: extension/runtime/src/main/java/io/quarkiverse/temporal/WorkflowStub.java (hunk 3 above)
Comment (rmanibus) [reply extension/runtime/src/main/java/io/quarkiverse/temporal/WorkflowStub.java]: done !

**Comment by rmanibus** [comment]
Ok I am satisfied with the current state, going to merge it for now !

**Comment by melloware** [comment]
sounds good!

**Pull request merged**